cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -66,10 +66,41 @@ type ClusterInterrogator interface {
 	RemoveMember(id uint64) ([]Member, error)
 }
 
+type etcdClient interface {
+	// MemberList lists the current cluster membership.
+	MemberList(ctx context.Context) (*clientv3.MemberListResponse, error)
+
+	// MemberAdd adds a new member into the cluster.
+	MemberAdd(ctx context.Context, peerAddrs []string) (*clientv3.MemberAddResponse, error)
+
+	// MemberAddAsLearner adds a new learner member into the cluster.
+	MemberAddAsLearner(ctx context.Context, peerAddrs []string) (*clientv3.MemberAddResponse, error)
+
+	// MemberRemove removes an existing member from the cluster.
+	MemberRemove(ctx context.Context, id uint64) (*clientv3.MemberRemoveResponse, error)
+
+	// MemberPromote promotes a member from raft learner (non-voting) to raft voting member.
+	MemberPromote(ctx context.Context, id uint64) (*clientv3.MemberPromoteResponse, error)
+
+	// Endpoints lists the registered endpoints for the client.
+	Endpoints() []string
+
+	// Sync synchronizes client's endpoints with the known endpoints from the etcd membership.
+	Sync(ctx context.Context) error
+
+	// Status gets the status of the endpoint.
+	Status(ctx context.Context, endpoint string) (*clientv3.StatusResponse, error)
+
+	// Close shuts down the client's etcd connections.
+	Close() error
```

Comment:
Can I ask how do you select the methods which should be implemented by the interface? i.e. why the `Close()` is here and `MemberUpdate` is not?

Reply:
Good question.

I selected the methods that _have_ to be in the interface, because they are called at least once by kubeadm. The `MemberUpdate` method is not called.

Reply:
perhaps we should just add all methods in the interface

Reply:
> perhaps we should just add all methods in the interface

I wouldn't want to include any methods not used by the "real" code.

The Go convention, as far as I understand it, is every consumer (e.g. every package) defines their own interface, and includes only the methods they call. There are exceptions to this convention, where the producer exports an Interface for consumers to import, but that's not the case here.

Reply:
ok, that's fair